artichoke-backend/src/extn/core/time/subsec.rs

```diff
@@ -0,0 +1,562 @@
+//! `Time.at` function call. These parameters (if specified) provide the number
+//! of subsecond parts to add, and a scale of those subsecond parts (millis, micros,
+//! and nanos).
+
+use crate::convert::implicitly_convert_to_int;
+use crate::extn::core::symbol::Symbol;
+use crate::extn::prelude::*;
+
+const NANOS_IN_SECOND: i64 = 1_000_000_000;
+
+const MILLIS_IN_NANO: i64 = 1_000_000;
+const MICROS_IN_NANO: i64 = 1_000;
+const NANOS_IN_NANO: i64 = 1;
+
+#[allow(clippy::cast_precision_loss)]
+const MIN_FLOAT_SECONDS: f64 = i64::MIN as f64;
+#[allow(clippy::cast_precision_loss)]
+const MAX_FLOAT_SECONDS: f64 = i64::MAX as f64;
+const MIN_FLOAT_NANOS: f64 = 0.0;
+#[allow(clippy::cast_precision_loss)]
+const MAX_FLOAT_NANOS: f64 = NANOS_IN_SECOND as f64;
+
+enum SubsecMultiplier {
+    Millis,
+    Micros,
+    Nanos,
+}
+
+impl SubsecMultiplier {
+    #[must_use]
+    const fn as_nanos(&self) -> i64 {
+        match self {
+            Self::Millis => MILLIS_IN_NANO,
+            Self::Micros => MICROS_IN_NANO,
+            Self::Nanos => NANOS_IN_NANO,
+        }
+    }
+}
+
+impl TryConvertMut<Option<Value>, SubsecMultiplier> for Artichoke {
+    type Error = Error;
+
+    fn try_convert_mut(&mut self, subsec_type: Option<Value>) -> Result<SubsecMultiplier, Self::Error> {
+        if let Some(mut subsec_type) = subsec_type {
+            let subsec_symbol = unsafe { Symbol::unbox_from_value(&mut subsec_type, self)? }.bytes(self);
+            match subsec_symbol {
+                b"milliseconds" => Ok(SubsecMultiplier::Millis),
+                b"usec" => Ok(SubsecMultiplier::Micros),
+                b"nsec" => Ok(SubsecMultiplier::Nanos),
+                _ => {
+                    let mut message = b"unexpected unit: ".to_vec();
+                    message.extend_from_slice(subsec_symbol);
+                    Err(ArgumentError::from(message).into())
+                }
+            }
+        } else {
+            Ok(SubsecMultiplier::Micros)
+        }
+    }
+}
+
+/// A struct that represents the adjustment needed to a `Time` based on a
+/// the parsing of optional Ruby Values. Seconds can require adjustment as a
+/// means for handling overflow of values. e.g. `1_001` millis can be requested
+/// which should result in 1 seconds, and `1_000_000` nanoseconds.
+///
+/// Note: Negative nanoseconds are not supported, thus any negative adjustment
+/// will generally result in at least -1 second, and the relevant positive
+/// amount of nanoseconds. e.g. `-1_000` microseconds should result in -1
+/// second, and `999_999_000` nanoseconds.
+#[derive(Debug, Copy, Clone)]
+pub struct Subsec {
+    secs: i64,
+    nanos: u32,
+}
+
+impl Subsec {
+    /// Returns a tuple of (seconds, nanoseconds). Subseconds are provided in
+    /// various accuracies, and can overflow. e.g. 1001 milliseconds, is 1
+    /// second, and `1_000_000` nanoseconds.
+    #[must_use]
+    pub fn to_tuple(&self) -> (i64, u32) {
```

Comment:
per discussion on Discord, is this better represented as `pub fn to_duration(&self) -> core::time::Duration`?

Reply:
One caveat here is the `Duration` type in `core` is unsigned.

Reply:
Filed:

- https://github.com/artichoke/artichoke/issues/2096